rope.c

```diff
@@ -211,6 +290,72 @@ static size_t count_bytes_in_utf8(const uint8_t *str, size_t num_chars) {
   return p - str;
 }
 
+/// Skip to the given UTF-8 offset in the string.
+/// \param str The string to search.
+/// \param chars The number of characters to skip.
+/// \param bytelen The length of the string in bytes.
+/// \return A pointer to the new offset in the string.
+static inline uint8_t *skip_utf8(uint8_t *str, size_t chars, size_t bytelen) {
+  size_t chars_seen = 0;
+  const uint8_t *const end = str + bytelen;
+  while (chars_seen < chars) {
+    size_t cp_size = codepoint_size(*str);
+    if (str + cp_size > end) {
+      break;
+    }
+    str += cp_size;
+    ++chars_seen;
+  }
+  return str;
+}
+
+/// Copy a number of UTF-8 characters into a new buffer. Does not write a null.
+/// \param dst The buffer to copy to.
+/// \param src The buffer to copy from.
+/// \param chars In: the number of characters to copy.
+///              Out: the number of characters written.
+/// \param max_size The maximum number of bytes to copy.
+/// \return The number of bytes written.
+static inline size_t copy_utf8(uint8_t *dst, const uint8_t *src, size_t *chars,
```

Comment:
This function is unnecessary given you know the exact number of bytes you want to copy. Use memcpy.

Reply:
This is slightly different than memcpy in that it copies up to but possibly less than a number of bytes and also counts characters while copying to avoid looping through the string twice.

Thinking about this a bit more, I think there are two cases here.
1. Copying the entire node - we know the exact byte and char length already, so I can change that path to use memcpy.
2. Copying only part of the node - this is where the number of bytes to copy depends on the char length. It seems that my options are to either count chars and then memcpy (2 loops, but only over a small string and memcpy is much better optimized), or keep this function for this case.

Is there another way to handle the second case that I'm missing?

Reply:
Mmm yeah I hear you about looping through the string twice, though I suspect the performance difference (if any) will be pretty small because the string is short and should fit in CPU cache. I think thats easier because we don't need your `copy_utf8` function but 🤷‍♀️